subsys/net/lib/nrf_cloud/src/nrf_cloud_jwt.c

```diff
@@ -8,12 +8,451 @@
+		LOG_ERR("Failed to create JWT JSON payload");
+		return NULL;
+	}
+
+	/* Convert payload JSON string to base64 URL */
+	pay_b64 = convert_str_to_b64_url(pay_str);
+
+	cJSON_free(pay_str);
+	pay_str = NULL;
+
+	if (!pay_b64) {
+		LOG_ERR("Failed to convert string to base64");
+		return NULL;
+	}
+
+	/* Create the base64 URL data to be signed */
+	unsigned_jwt = jwt_header_payload_combine(JWT_ES256_HDR, pay_b64);
+
+	nrf_cloud_free(pay_b64);
+	pay_b64 = NULL;
+
+	return unsigned_jwt;
+}
+
+static int jwt_signature_get(const int sec_tag, const char *const jwt,
+			     char *const sig_buf, size_t sig_sz)
+{
+	int err;
+	size_t o_len;
+	char der_buf[PRV_KEY_SZ];
+	uint8_t sig_raw[ECDSA_SHA_256_SIG_SZ];
+
+	/* Get the key which will be used to sign the JWT */
+	err = get_key_from_cred(sec_tag, der_buf);
+	if (err) {
+		LOG_ERR("Failed to get private key material, error: %d", err);
+		return -EIDRM;
+	}
+
+	err = sign_message_der(der_buf, jwt, sig_raw);
+	if (err) {
+		LOG_ERR("Failed to sign JWT, error: %d", err);
+		return -ENOEXEC;
+	}
+
+	LOG_HEXDUMP_DBG(sig_raw, sizeof(sig_raw), "Signature: ");
+
+	/* Convert signature to base64 URL */
+	err = base64_encode(sig_buf, sig_sz, &o_len, sig_raw, sizeof(sig_raw));
+	if (err) {
+		LOG_ERR("base64_encode failed, error: %d", err);
+		return -EIO;
+	}
+
+	base64_url_format(sig_buf);
+	return 0;
+}
+
+static int jwt_signature_append(const char *const unsigned_jwt, const char *const sig,
+				char *const jwt_buf, size_t jwt_sz)
+{
+	int err = 0;
+
+	/* Get the size of the final, signed JWT: +1 for '.' and null-terminator */
+	size_t final_sz = strlen(unsigned_jwt) + 1 + strlen(sig) + 1;
+
+	if (final_sz > jwt_sz) {
+		/* Provided buffer is too small */
+		err = -E2BIG;
+	}
+
+	/* JWT final form:
+	 * <base64_header>.<base64_payload>.<base64_signature>
+	 */
+	if (err == 0) {
+		int ret = snprintk(jwt_buf, jwt_sz, "%s.%s", unsigned_jwt, sig);
+
+		if ((ret < 0) || (ret >= jwt_sz)) {
+			err = -ETXTBSY;
+		}
+	}
+
+	return err;
+}
+
+static int custom_jwt_generate(struct jwt_data *const jwt)
+{
+	int err = 0;
+	char *unsigned_jwt;
+	uint8_t jwt_sig[B64_SIG_SZ];
+
+	/* Create the JWT to be signed */
+	unsigned_jwt = unsigned_jwt_create(jwt);
+	if (!unsigned_jwt) {
+		LOG_ERR("Failed to create JWT to be signed");
+		return -EIO;
+	}
+
+	LOG_DBG("Message to sign: %s", unsigned_jwt);
+
+	/* Get the signature of the unsigned JWT */
+	err = jwt_signature_get(jwt->sec_tag, unsigned_jwt, jwt_sig, sizeof(jwt_sig));
+	if (err) {
+		LOG_ERR("Failed to get JWT signature, error: %d", err);
+		nrf_cloud_free(unsigned_jwt);
+		return -ENOEXEC;
+	}
+
+	/* Append the signature, creating the complete JWT */
+	err = jwt_signature_append(unsigned_jwt, jwt_sig, jwt->jwt_buf, jwt->jwt_sz);
+
+	nrf_cloud_free(unsigned_jwt);
+
+	if (!err) {
+		LOG_DBG("JWT:\n%s", jwt->jwt_buf);
+	} else {
+		LOG_ERR("Failed to append JWT signature, error: %d", err);
+	}
+
+	return err;
+}
+#endif /* CONFIG_NRF_CLOUD_JWT_SOURCE_CUSTOM */
```

Comment:
Is all of this original to this PR, or brought in from somewhere else in the SDK?

Reply:
it is new

Reply:
Got it. I'll give it a more detailed review, then, unless you'd prefer I hold off for the next revision

Reply:
please review.
i have no major revisions planned, except for changing the overlay to align with https://github.com/nrfconnect/sdk-nrf/pull/13841.